### PR TITLE
[SCB-685] Service comb chassis must support default values in query parameters

### PR DESCRIPTION
**Pull request by acsukesh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java [hunk 1]
```diff
@@ -43,6 +54,12 @@ public Object getValue(HttpServletRequest request) throws Exception {
         value = request.getParameterValues(paramPath);
       } else {
         value = request.getParameter(paramPath);
+        if (value == null || value.equals("")) {
```
Comment (wujimin) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: 1."" should not change to default value?
2.only process query parameter? what about header/cookie and so on?

File: samples/springmvc-sample/springmvc-consumer/src/main/java/org/apache/servicecomb/samples/springmvc/consumer/SpringmvcConsumerMain.java [hunk 2]
```diff
@@ -51,9 +51,12 @@ public static void main(String[] args) throws Exception {
     // RestTemplate Consumer
     String sayHiResult =
         restTemplate.postForObject("cse://springmvc/springmvchello/sayhi?name=Java Chassis", null, String.class);
+    String sayHiDefaultResult =
```
Comment (liubao68) [inline samples/springmvc-sample/springmvc-consumer/src/main/java/org/apache/servicecomb/samples/springmvc/consumer/SpringmvcConsumerMain.java]: It's better to add test cases in demo projects. Sample are short examples, and demos folder are integration tests, you can add as many test cases you like. 

And maybe you need to test some primitive types default value, and from the following code, they seems not properly handled.

File: swagger/swagger-generator/generator-core/src/main/java/org/apache/servicecomb/swagger/generator/core/processor/parameter/AbstractParameterProcessor.java [hunk 3]
```diff
@@ -55,6 +57,18 @@ protected void setParameterName(Object annotation, OperationGenerator operationG
...
 
+  protected void setParameterDefaultValue(Object annotation, T parameter) {
+    String defaultValue = getAnnotationParameterDefaultValue(annotation);
+    if (StringUtils.isNotEmpty(defaultValue)) {
+      parameter.setDefaultValue(defaultValue);
+    }
+
+  }
+
+  protected String getAnnotationParameterDefaultValue(Object annotation) {
```
Comment (liubao68) [inline swagger/swagger-generator/generator-core/src/main/java/org/apache/servicecomb/swagger/generator/core/processor/parameter/AbstractParameterProcessor.java]: primitive and other types default value not properly handled?

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java [hunk 4]
```diff
@@ -32,4 +33,13 @@ protected QueryParameter createParameter() {
   protected String getAnnotationParameterName(Object annotation) {
     return ((RequestParam) annotation).name();
   }
+
+  @Override
+  protected String getAnnotationParameterDefaultValue(Object annotation) {
+    String defaultValue = ((RequestParam) annotation).defaultValue();
+    if (defaultValue.equals(ValueConstants.DEFAULT_NONE)) {
+      return "";
```
Comment (liubao68) [inline swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java]: maybe null is better?

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java (hunk 1 above)
Comment (liubao68) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: Maybe we can add defaultValue to base interface

Comment (acsukesh) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: "" comes only when user inputs the same and is considered as valid input parameter.

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java (hunk 4 above)
Comment (acsukesh) [reply swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java]: might cause NullPointerException, also some message might get added with null which may not be correct , so returning empty

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java (hunk 1 above)
Comment (liubao68) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/codec/param/QueryProcessorCreator.java]: OK. For  http://10.57.65.159:8080/codeFirstSpringmvc/reduce?a=
give default Value is better.

File: swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java (hunk 4 above)
Comment (liubao68) [reply swagger/swagger-generator/generator-springmvc/src/main/java/org/apache/servicecomb/swagger/generator/springmvc/processor/annotation/RequestParamAnnotationProcessor.java]: Maybe it's better to return ValueConstants.DEFAULT_NONE. When no default value specified and users do not give the parameter value, we can throw an BadRequest exception.

**Comment by liubao68** [comment]
I'll merge this PR. Following work need to be done:
1. Support JAX-RS
2. Adding more test cases, to cover default values not specified and user do not parse any value. See review comments: " Maybe it's better to return ValueConstants.DEFAULT_NONE. When no default value specified and users do not give the parameter value, we can throw an BadRequest exception or give default value based on parameter type, e.g. int is 0 and String is null."

**Pull request merged**